helix-core/src/main/java/org/apache/helix/controller/stages/TaskGarbageCollectionStage.java

```diff
@@ -23,34 +27,53 @@ public AsyncWorkerType getAsyncWorkerType() {
   }
 
   @Override
-  public void execute(ClusterEvent event) {
+  public void process(ClusterEvent event) throws Exception {
     WorkflowControllerDataProvider dataProvider =
         event.getAttribute(AttributeName.ControllerDataProvider.name());
+    event.addAttribute(AttributeName.WORKFLOW_CONFIG_MAP.name(),
+        dataProvider.getWorkflowConfigMap());
+    event.addAttribute(AttributeName.RESOURCE_CONTEXT_MAP.name(), dataProvider.getContexts());
+
+    super.process(event);
+  }
+
+  @Override
+  public void execute(ClusterEvent event) {
+    Map<String, WorkflowConfig> workflowConfigMap =
+        event.getAttribute(AttributeName.WORKFLOW_CONFIG_MAP.name());
+    Map<String, ZNRecord> resourceContextMap =
+        event.getAttribute(AttributeName.RESOURCE_CONTEXT_MAP.name());
     HelixManager manager = event.getAttribute(AttributeName.helixmanager.name());
 
-    if (dataProvider == null || manager == null) {
+    if (manager == null) {
       LOG.warn(
-          "ResourceControllerDataProvider or HelixManager is null for event {}({}) in cluster {}. Skip TaskGarbageCollectionStage.",
+          " HelixManager is null for event {}({}) in cluster {}. Skip TaskGarbageCollectionStage.",
           event.getEventId(), event.getEventType(), event.getClusterName());
       return;
     }
 
-    Set<WorkflowConfig> existingWorkflows =
-        new HashSet<>(dataProvider.getWorkflowConfigMap().values());
+    Set<WorkflowConfig> existingWorkflows = new HashSet<>(workflowConfigMap.values());
```

Comment:
If we assumed soft copy was okay from above, then why are we creating a new set? Also this is not a hard copy either is it?

Reply:
Outdated